### ISSUE TITLE
Pi-Hole breaks Spotify Desktop in Arch Linux. Tried whitelisting to no avail.

### ISSUE THREAD

**Issue by zimmertr** [pi-hole/pi-hole#364]

**Comment by dschaper** [comment]
Thanks, which Pi are you using? I'll try to duplicate it with a test configuration here. Latest install of ALARM on a Pi2B? And how are you installing Spotify, through the AUR?

**Comment by zimmertr** [comment]
The Pi is a Raspberry Pi 1 Model B. It's running Raspian 8 Jessie. Pi-Hole is version 2.5.3 and the WebUI is 1.1.4. Arch Linux has the most currently available packages available in the user repository system wide. I'm using Spotify-Beta 1.0.23.93-3 from the AUR found here: https://aur.archlinux.org/packages/spotify-beta/?comments=all

I have whitelisted all of the following domains which were aggregated from both the pi-hole log and the whitelist chrome extension: 

gucl-6.ap.spotify.com
appresolve.spotify.com
ash2-accesspoint-a86.ap.spotify.com
ash2-accesspoint-a82.ap.spotify.com
ash2-accesspoint-a29.ap.spotify.com
play.spotify.com
play.spotify.edgekey.net
glue-static.s3-external-3.amazonaws.com
connect.facebook.net
dwbsggjlqw.spotilocal.com
www.facebook.com
staticxx.facebook.com
static.xx.fbcdn.net
embed.spotify.com
d2b1xqaw2ss8na.cloudfront.net
d2j0f09xqmv212.cloudfront.net
api-read.facebook.com
graph.facebook.com
ssl.google-analytics.com
www.googletagservices.com
ajax.googleapis.com

Starting Spotify makes the following domains appear in the log despite being whitelisted: 
apresolve.spotify.com
gucl-6.ap.spotify.com
ash2-accesspoint-a69.ap.spotify.com******

*******The two numbers in this domain change to something new everytime I open a new instance of spotify. 69, 86, 82, 29, etc. I tried to whitelist with * wildcards but it didn't work and said it was an invalid domain. 

I'm not sure what ALARM is so I can't tell you that information. I've only had this pi for a few days and I'm a noob.

**Comment by dschaper** [comment]
Thanks for the information. ALARM is just the shorthand for Arch Linux ARM (Arch on RasPi)... We currently don't support wildcards in the whitelists, but that's something that is on the project list. It will take me just a little bit of time to spin up an Arch install, then I'll see what I can find.

**Comment by dschaper** [comment]
Sorry, re-reading the post. You're running Raspbian but you've installed an Arch package?

**Comment by zimmertr** [comment]
I'm using Arch Linux on a desktop computer at 192.168.1.3 and the RasPi has Rasbian on it located at 192.168.1.2. Starting Spotify on Arch Linux and using the RasPi as the DNS server causes the issue. Sorry for the confusion.

**Comment by dschaper** [comment]
Ah, okay, that's a bit easier to set up then, no need for ALARM. Thanks!

**Comment by dschaper** [comment]
It will take just a little bit while I set up an Arch install...

**Comment by zimmertr** [comment]
No rush, I'm off to the bar to play pool with the fellas anyway.

**Comment by PromoFaux** [comment]
@zimmertr Are you running with the default block lists? Because there are no spotify domains in any of those lists...

```bash
fap@fappotron:~$ for list in /etc/pihole/list.*;do echo $list;grep 'spotify' $list;done
/etc/pihole/list.0.raw.githubusercontent.com.domains
/etc/pihole/list.1.adblock.gjtech.net.domains
/etc/pihole/list.2.mirror1.malwaredomains.com.domains
/etc/pihole/list.3.sysctl.org.domains
/etc/pihole/list.4.zeustracker.abuse.ch.domains
/etc/pihole/list.5.s3.amazonaws.com.domains
/etc/pihole/list.6.s3.amazonaws.com.domains
/etc/pihole/list.7.hosts-file.net.domains
/etc/pihole/list.8.raw.githubusercontent.com.domains
```

I've just tried the windows version of spotify and it works fine, so this seems like it may either something unique to Arch linux spotify, or you're running additional blocklists...?

**Comment by PromoFaux** [comment]
Also to note, you might want to reset your whitelist, as some of those domains are tracking domains that will now be tracking you  across the internet :)

**Comment by zimmertr** [comment]
I assume I'm using the default list as I simply installed Pi-Hole and let it do it's thing. I never did anything extra aside from installing the WebUI.

Can I simply reset the whitelist by editing a file somewhere? Is there a more elegant solution implemented?

**Comment by PromoFaux** [comment]
Easiest thing to do would be to clear out `/etc/pihole/whitelist.txt` and then run `gravity.sh`

(either `rm /etc/pihole/whitelist.txt` or `echo "" > /etc/pihole/whitelist.txt`)

**Comment by zimmertr** [comment]
Here is the resulting output: http://pastebin.com/fnQ6dA5B

Spotify is still not working, in case you were wondering.

**Comment by PromoFaux** [comment]
OK, so your whitelist is reset. that's good. I wasn't expecting spotify to work after that, but at least you don't have trackers whitelisted. I have been out this afternoon, but I'm about to install ARCH and see what I can work out.

**Comment by dschaper** [comment]
Okay, I have my VM set up with Arch, noticed the AUR package for Spotify-Beta is flagged as out of date? (I'm still going to use the AUR version and see what I can find for blocking...)

**Comment by zimmertr** [comment]
That's weird that spotify-beta marked as out of date. It was last updated 1 week and 5 days ago. And it is one version number ahead of Spotify in the other repo.

**Comment by dschaper** [comment]
Well as long as we are both using the same AUR PGKBUILD then we should be able to duplicate the issue... Looks like it was just flagged, but no comments left on the update?

**Comment by dschaper** [comment]
And I'm just installing Spotify-Beta via Yaourt and specifying defaults for everything on my install...

**Comment by dschaper** [comment]
I'm not able to replicate the issue, if you start Spotify desktop from a terminal window with just `spotify` and leave the terminal open, do any errors show in the terminal? I'm able to synchronize offline playlists and stream audio. The only hit that I'm seeing that Pi-hole is blocking is the `www.googletagservices.com` but that shouldn't keep you in offline mode...

**Comment by zimmertr** [comment]
Wow, well it's unfortunate that you went through all of this trouble then.

I do quite a few errors: http://pastebin.com/b16bARnr

**Comment by dschaper** [comment]
```22:55:15.491 E [offline_key_store.cpp:85        ] Offline: failed to load key stores, error: sync_error_offline_key_store_file_not_found (7007), path: /home/dan/.cache/spotify/Users/danjschaper-user/cached
``` I do see this error...

**Comment by zimmertr** [comment]
The last bit of the output that I posted as the same reoccurring error. Grep the log for "AP Socket Error: Timeout Reached" to see the portions I'm talking about

**Comment by dschaper** [comment]
Okay, I see those errors. On the Pi-hole, can you open a terminal and run `sudo tail -f /var/log/pihole.log | grep gravity.list` for me? That command will list all the domains that are being redirected to the Pi-hole, and if we see the Spotify domains coming up, then we'll be closer to finding out if the Pi-hole is the one causing the errors...

**Comment by dschaper** [comment]
I picked one of the domains in those errors and I'm able to ping them without the Pi-Hole intercepting:

```
[dan@localhost ~]$ ping guc1-6.ap.spotify.com
PING guc1.gcplb.gslb.spotify.com (104.154.127.99) 56(84) bytes of data.
64 bytes from 99.127.154.104.bc.googleusercontent.com (104.154.127.99): icmp_seq=1 ttl=36 time=61.5 ms
64 bytes from 99.127.154.104.bc.googleusercontent.com (104.154.127.99): icmp_seq=2 ttl=36 time=61.1 ms
64 bytes from 99.127.154.104.bc.googleusercontent.com (104.154.127.99): icmp_seq=3 ttl=36 time=60.9 ms
64 bytes from 99.127.154.104.bc.googleusercontent.com (104.154.127.99): icmp_seq=4 ttl=36 time=61.9 ms
```

**Comment by zimmertr** [comment]
Interesting. I can successfully ping that domain as well. Doesn't the -f flag cause tail to follow the output of the file indefinitely? Anyway, it's not returning any output and it's been running for over 5 minutes now.

**Comment by dschaper** [comment]
Yes, the -f causes the output to scroll. If you're not seeing any hits, then Pi-hole isn't intercepting anything. If you do `sudo tail -f /var/log/pihole.log` by itself without the `grep` you should see all the domains that are being resolved, both the redirected and the allowed traffic. You can do that to make sure that your installs are using the Pi-hole as their resolver...

**Comment by dschaper** [comment]
Can I close out this issue, it looks like it's unrelated to the Pi-hole?

**Comment by zimmertr** [comment]
Feel free. Do you have any suggestions on how I can remedy the issue? It's so strange to me that Spotify works fine if I use any other DNS server. 

Could it possibly be a setting on my OpenWRT flashed router? I'm at a loss.

**Comment by dschaper** [comment]
I would guess it's possible, but I'm running OpenWRT on my install and had no problems. Did you check the `/etc/resolv.conf` and make sure the both the Arch and the Pi are behaving?

**Comment by zimmertr** [comment]
I'm unsure how to go about what you asked. Or which linux device I'm supposed to check for that information since the Pi, my router, and my computer all have that file.

I'm assuming my computer. Which lists both the proper IP address as well as the mac address of the pi that it is using for pi-hole.

**Comment by dschaper** [comment]
Do a `cat /etc/resolv.conf` on the Arch box, and the same command on the Pi-hole. Then post up the results and let's take a look and see. If you can do that both with the Pi-hole configured as the DNS server, and with the Pi-hole taken out, (So the setting when Spotify works and the settings when Spotify breaks,)

**Comment by zimmertr** [comment]
From Arch, with Pi-Hole: http://pastebin.com/47RAACvK
From Pi-Hole: http://pastebin.com/75UpN7ST
From Arch, with 8.8.8.8: http://pastebin.com/zRFBW51v

**Comment by dschaper** [comment]
Okay, lots of questions on this one...

That IPv6 address duplication kind of concerns me. Is the Pi-Hole at 192.168.1.2? And are you setting the DNS addresses in OpenWRT, or are you setting them directly. Also do you have IPv6 blocking enabled? 

Can you post up an `ip a` for the Pi-hole?

**Comment by dschaper** [comment]
Which router are you using? I have the below setup:

![capture](https://cloud.githubusercontent.com/assets/2782850/13557032/92259eec-e39c-11e5-9846-e153d484e63b.PNG)


(Closing the issue, but we can still discuss the problems and see if we can't figure it out.)

**Comment by zimmertr** [comment]
Router: 192.168.1.1
Pi-Hole: 192.168.1.2
Arch Linux Box: 192.168.1.3
Miscellaneous other computers 192.168.1.4-9

The DNS addresses have been set both in each application within my intranet as well as in OpenWRT.  Here is a screenshot of me setting it in OpenWRT: http://i.imgur.com/ImWsgwF.png

Here is a screenshot of my router information: http://imgur.com/N2lEW6T

I do not believe that I have IPV6 blocking enabled. I have no experience with IPV6 and thus have left all of the settings as their default values in OpenWRT. If I can provide you with a screenshot or information please let me know. 

Here is a copy of $ip a being run from the Pi: http://pastebin.com/1kmjpLjv

**Comment by dschaper** [comment]
I just can't seem to replicate it. Can you use a non-Premium Spotify account and see if it's related to that aspect? I just have a standard account and not a Premium.

Here's my start-up
http://pastebin.com/1WbNuQmA

**Comment by zimmertr** [comment]
I made a new account for the free version and that didn't fix the issue either. =[

Here is a copy of my startup as well: http://pastebin.com/a4DD6Crp

**Comment by zimmertr** [comment]
For what it's worth, I just discovered that I can't connect to my wifi using my laptop and the dns server either. 

My laptop also has Arch on it. And uses WICD to connect to the wifi. When I specify the Pi-Hole as the DNS server it connects successfully to the network but it has DNS issues. I can ping 8.8.8.8 but I can't ping google.com. If I set 8.8.8.8 as the DNS server I can get on the internet, though. Do you think there's an issue with my router configuration?

**Comment by dschaper** [comment]
It's beginning to sound like there might be. What router do you have? I noticed the firmware was the older CC instead of the newer DD trunk build. We'll keep at this until we figure out what's what! :)

**Comment by zimmertr** [comment]
Sorry for the delay in my response. I have been very busy with my new job. It's a Netgear wndr3800.

**Comment by Bruut8** [comment]
Signing in to report the same error on my setup:

Netgear r6250 router,  stock firmware 
Pi2 with dietpi running only pihole as dns
Pc, Windows 7 with spotify desktop

When setting the dns on the router to the internal pi adress,  everything runs great,  less adds! BUT spotify stops working. 
The moment I let the router use the standard dns servers everything is  fine.. 

I'd love to debug but will not be able to in the coming month. You're not alone though!

**Comment by dschaper** [comment]
Okay, keep the issue in mind, and when you are at a place where you can help us debug we can re-open the issue and take a look...

**Comment by zimmertr** [comment]
After my Pi stopped working I decided to install debian in a virtual machine on my ESXi host. From there I installed pi-hole and configured it to be used as my DNS. Spotify is now in working order on all machines.

It appears that the issue was either the raspberry pi itself or a configuration of pi-hole on the device. Because it works fine from the VM.

**Comment by zimmertr** [comment]
Never mind, after a reboot they're not working anymore. They must have have been working because the client wasn't restarted after setting the DNS server.

**Comment by Bruut8** [comment]
Back from vacation and reporting in. Pi had been running for 2 months straight, no settings changed since last update.

The spotify issue has resolved itself it seems. The desktop client, web client and mobile client all work as intended. 

Thanks for listening!

**Comment by snaggyheadshot** [comment]
I'm having the Spotify issue, but weirdly only with searching songs and loading public playlists. So i'm fine to play playlistst I created by myself but I am unable to load public playlists and searching. It simply says "an error occurred". But is does work when I change my DNS to the Google DNS servers.

I do only have this problem on my iDevices. Mac and Windows works fine.

**Comment by j0hanSE** [comment]
Just as the person above me, my girlfriend was complaining about her Spotify app didnt work, and i had the same problem aswell. I cant search or public playlists. I'm on Pi-hole Version v2.7.1.

How to solve it?

**Comment by snaggyheadshot** [comment]
@j0hanSE Solved it for me as well! Thanks. Pi Hole should really give a option to whitelist all these things in once. I had to whitelist a lot of websites to be able to use for example Google shopping. Ofcourse these are ads, but ads where I did ask for.

**Comment by dschaper** [comment]
If you are using the CLI, you should be able to run `pihole -w domain.com domain2.com domain3.com` etc, with all the domains listed on one line. Let us know if that's not the case.

**Comment by snaggyheadshot** [comment]
Well. If I whitelist the domain it still does not allow me to click on affiliate/tracking links. I can understand people who use Pi-Hole don't wat to be tracked, but for me are product comparing sites very helpful. And besides that I also can't visit sites who have tracker links in order to view a video of for example my national news channel.

**Comment by Zegnat** [comment]
> Pi Hole should really give a option to whitelist all these things in once.

Pi-Hole works the other way around: you get the option what to blacklist.

By default it will load several lists of domains and combine them into a single blacklist (see [adlists.default](https://github.com/pi-hole/pi-hole/blob/master/adlists.default)). Already there are some lists that are not included by default because they block too much (e.g. the Mahakala list) or because they break certain functionality (e.g. the Windows 10 telemetry list).

Maybe some of the lists that are enabled by default do not work for you, and you should disable those. That isn't really bulk-whitelisting though, it is choosing to not blacklist something.

>  If I whitelist the domain it still does not allow me to click on affiliate/tracking links.

You could file a feature request on [packtloss/pihole-extension](https://github.com/packtloss/pihole-extension), as it could have Chrome scan for all domain names used on a specific page and whitelist all of them in one go. That might be closer to what you are looking for. It is not feasible (I would think) for Pi-Hole to scrape an entire domain to auto-whitelist dependencies the way you seem to suggest.

**Comment by snaggyheadshot** [comment]
It also seems to block login on several sites. I'm unable to login at alibaba.com for example. I will check to exclude some lists. Thanks.

**Comment by zimmertr** [comment]
I can't use the Concerts feature of Browse now. Still on Arch using the Debian VM/Pi-Hole I made a while ago in this thread. Can reproduce on my MacBook running Yosemite.

**Comment by dschaper** [comment]
@snaggyheadshot Could you open a new issue for that for us? I'd like to keep this issue for Spotify related traffic. 

Thanks!

**Comment by dschaper** [comment]
@zimmertr I'll see if I can hack around a little more and shake something loose. This is still only with Spotify Premium, correct? Or does this affect regular Spotify as well?

**Comment by zimmertr** [comment]
Tested on premium using the desktop client. Cant reproduce in the browser since the functionality doesn't exist it appears.

**Comment by zczo** [comment]
I have added spclient.wg.spotify.com to the whitelist and it's working for me.

**Comment by luisdoubrava** [comment]
Any update on this? I also added spclient.wg.spotify.com but IOS still can't connect. Desktop works fine.

**Comment by zimmertr** [comment]
I no longer experience this issue on any platform.

**Comment by KaffineAddict** [comment]
@luisdoubrava I and a friend also have issues with spotify iOS version but there isnt anything in the blocked domain logs.

**Comment by JamborJan** [comment]
Same issue here:
- Spotify Premium
- iOS (latest)
- Always offline
- Desktop okay
- no obvious logs in pi-hole

Should we open a new issue or do you reopen this one?

**Comment by dschaper** [comment]
If none of the fixes in this ticket are helpful, then it would be good to open a new issues ticket referencing that this is an iOS related problem so we can work it from there.

**Comment by dokape** [comment]
pihole 3.0.1: 
Spotify 8.4.3.884 on iOS 10.3.2 (maybe other/older versions affected too)  did not work correctly:    SEARCH was not available.
On Android the problem did not exist!

Putting "spclient.wg.spotify.com" to the whitelist did spotify work correct. 
So there is still an issue with blocking domains from spotify. Event if there are explicit no blocking,
some spotify-domains seems unreachable when using pihole as dns-server.

I had this problem in Juli 2016 before, solved it. After a complete new install of pihole with version 3 a few days ago the problem appeared again. 
So the problem is still actual.

**Comment by dschaper** [comment]
This thread is for the Arch Linux platform, please open a new ticket for iOS issues.

Thank you.
